hypothesis-python/src/hypothesis/extra/array_api.py

```diff
@@ -0,0 +1,919 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Most of this work is copyright (C) 2013-2021 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+import math
+import sys
+from collections import defaultdict
+from numbers import Real
+from types import SimpleNamespace
+from typing import (
+    Any,
+    Iterable,
+    Iterator,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+from warnings import warn
+
+from hypothesis import strategies as st
+from hypothesis.errors import HypothesisWarning, InvalidArgument
+from hypothesis.extra._array_helpers import (
+    NDIM_MAX,
+    BasicIndex,
+    BasicIndexStrategy,
+    BroadcastableShapes,
+    Shape,
+    array_shapes,
+    broadcastable_shapes,
+    check_argument,
+    check_valid_dims,
+    mutually_broadcastable_shapes as _mutually_broadcastable_shapes,
+    order_check,
+    valid_tuple_axes as _valid_tuple_axes,
+)
+from hypothesis.internal.conjecture import utils as cu
+from hypothesis.internal.coverage import check_function
+from hypothesis.internal.reflection import proxies
+from hypothesis.internal.validation import (
+    check_type,
+    check_valid_bound,
+    check_valid_integer,
+    check_valid_interval,
+)
+from hypothesis.strategies._internal.strategies import check_strategy
+from hypothesis.strategies._internal.utils import defines_strategy
+
+__all__ = [
+    "make_strategies_namespace",
+]
+
+
+INT_NAMES = ["int8", "int16", "int32", "int64"]
+UINT_NAMES = ["uint8", "uint16", "uint32", "uint64"]
+ALL_INT_NAMES = INT_NAMES + UINT_NAMES
+FLOAT_NAMES = ["float32", "float64"]
+NUMERIC_NAMES = ALL_INT_NAMES + FLOAT_NAMES
+DTYPE_NAMES = ["bool"] + NUMERIC_NAMES
+
+DataType = TypeVar("DataType")
+
+
+class PrettyArrayModule:
+    """Wrapper for array modules so that they may have nice reprs."""
```

Comment:
I don't think we ever show the repr of array modules anymore, so we can safely remove this layer of indirection and get a few stackframes back 🙂

Reply:
Ah so I left this in because I wasn't sure if all Array API implementations will have the `__name__` to use for error/warning messages.

- [ ] I need to ask someone about this